### ISSUE TITLE
How do I set the timestamp on files in a zip file I create?

### ISSUE THREAD

**Issue by russellsilva** [rubyzip/rubyzip#316]
I'd like to set the timestamp on files in a zip file to something other than the current date and time. I've found two potential ways to do this in the docs, but neither work. The first just doesn't work; the timestamp isn't changed. The second results in an exception: `rubyzip-1.2.0/lib/zip/entry.rb:464:in '==': undefined method 'dos_equals' for 2017-01-28 14:03:29 -0800:Time (NoMethodError)`.

Below are my test cases demonstrating my failed attempts. ruby 2.3.1, rubyzip 1.2.0.

[Class: Zip::File — Documentation for rubyzip/rubyzip (master)](http://www.rubydoc.info/github/rubyzip/rubyzip/Zip/File#get_output_stream-instance_method)

```
$ cat create-file-with-mtime.rb
require 'zip'

io = Zip::File.open('output/foo.zip', Zip::File::CREATE);
io.get_output_stream("foo.txt", time: Time.now - (10 * 86400)) do |f|
        f.puts("Hello, World!")
end
io.close
$ bundle exec ruby create-file-with-mtime.rb
$ zipinfo output/foo.zip
Archive:  output/foo.zip   128 bytes   1 file
-rw-r--r--  5.2 unx       14 t- defN  7-Feb-17 14:02 foo.txt
1 file, 14 bytes uncompressed, 16 bytes compressed:  -14.3%
```

[Method: Zip::Entry#time= — Documentation for rubyzip/rubyzip (master)](http://www.rubydoc.info/github/rubyzip/rubyzip/Zip%2FEntry:time=)

```
$ cat set-file-mtime.rb
require 'zip'

Zip::File.open('output/foo.zip') do |zf|
  zf.entries.each do |entry|
    entry.time = (Time.now - (10 * 86400))
  end
end
$ bundle exec ruby set-file-mtime.rb
/Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/entry.rb:464:in `==': undefined method `dos_equals' for 2017-01-28 14:03:29 -0800:Time (NoMethodError)
        from /Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/entry_set.rb:53:in `=='
        from /Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/entry_set.rb:53:in `=='
        from /Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/file.rb:343:in `!='
        from /Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/file.rb:343:in `commit_required?'
        from /Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/file.rb:309:in `commit'
        from /Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/file.rb:334:in `close'
        from /Users/rsilva/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/rubyzip-1.2.0/lib/zip/file.rb:104:in `open'
        from set-file-mtime.rb:3:in `<main>'
```

If I replace `entry.time = (Time.now - (10 * 86400))` with `entry.time = ::Zip::DOSTime.at(Time.now - (10 * 86400))`, the exception doesn't happen, but the timestamp doesn't change, either. I don't think `::Zip::DOSTime` was meant as part of the user interface for `rubyzip` anyway.

Is there a way to set the timestamp on files in a zip file using rubyzip?

**Comment by seth-at-at** [comment]
Looking at it no, DOSTime is only there to create its own timestamp, and the way it's written it looks like if you wanted to set your own entry time you'd have to find a way to turn the DOSTime methods off.

**Comment by marisveide** [comment]
Hi!

Is it realistic to get the answer on this?

Here's how far I made it work:
```ruby
def save_to_zip(file_path)
  Zip::OutputStream.open(file_path) do |out|
    @sheets.each do |csv|
      name = csv.name
      extra = time_for_zip
      out.put_next_entry("#{name}.#{@file_extension}", nil, extra)
      tmpfile = csv.tmpfile
      tmpfile.close
      source = File.open(tmpfile.path, 'r')
      source.each do |line|
        out.write(line)
      end
    end
  end
end

def time_for_zip
  return nil if @time_zone.blank?

  timestamp = Zip::ExtraField.new('UT\x5\0\x3\250$\r@Ux\0\0')
  localtime_str = Time.now.in_time_zone(@time_zone).strftime("%Y-%m-%dT%H:%M:%S")
  dos_time_in_store_tz = ::Zip::DOSTime.parse(localtime_str)

  timestamp['UniversalTime'].ctime = dos_time_in_store_tz
  timestamp['UniversalTime'].atime = dos_time_in_store_tz
  timestamp['UniversalTime'].mtime = dos_time_in_store_tz

  timestamp
end
```

I have been banging my head the whole day on this, maybe you can help? I am zipping the file with RubyZip and I need to set the time of that file creation/update/modification to the certain time in time zone (which depends on the client time zone I have in the @time_zone variable).

I know it is most likely super incorrect, and I have taken that magic string 'UT\x5\0\x3\250$\r@Ux\0\0' from the RubyZip tests file, and I have no idea what this is. LOL. However - I have made it work now on my PC. It really zips the file and sets the correct timestamps for it according to the specified time zone.

BUT - it doesn't work on the app server, which as OS time zone is in UTC time zone. It generates some other time for files which doesn't match.

Maybe at least we can put our brains together to make sense of it?

HELP!
Anybody?

**Comment by marisveide** [comment]
Solved like described here:
https://github.com/rubyzip/rubyzip/pull/40

**Comment by hainesr** [comment]
I think if you use the option `:restore_times` in `Zip::File` now that we've merged #481 this should allow you to set whatever time you want and have it preserved. Let me know if not.
